x/evm/types/msg.go

```diff
@@ -314,9 +330,26 @@ func (msg MsgEthereumTx) AsTransaction() *ethtypes.Transaction {
 	return ethtypes.NewTx(txData.AsEthereumData())
 }
 
-// AsMessage creates an Ethereum core.Message from the msg fields
+// AsMessage creates an Ethereum core.Message from the msg fields. If the msg is impersonated, ie
+// has non-empty OnBehalf, then From is replaced with it. Otherwise, the from address is derived
+// from the signature.
 func (msg MsgEthereumTx) AsMessage(signer ethtypes.Signer, baseFee *big.Int) (core.Message, error) {
```

Comment:
this method is important as it replaces `from` in core eth msg. it should be used everywhere instead of `ethtypes.Transaction.AsMessage()`.